### PR TITLE
JWT tokens & auth

### LINKED ISSUE #58: Implement JWT

### PR DESCRIPTION
**Pull request by arkus7**
Closes #58 

**Checklist:**
- [x] Create an access token
- [x] Create a refresh token
- [x] Return access and refresh tokens on the login 
- [x] Add JWT Middleware for user authentication
- [x] Create refresh token endpoint
- [x] Validate expiration of tokens
- [x] Invalidate access tokens
- [x] Update logout endpoint
- [x] Authorization tests
- [x] Documentation 

**Known issues:**
- `Imperial` controller is not working anymore. It still needs to be refactored as we don't serve frontend in the app, therefore I won't update it right now.

### CODE REVIEW

File: Sources/App/Controllers/PostController.swift [hunk 1]
```diff
@@ -44,14 +44,13 @@ class PostController: RouteCollection {
     }
     
     func postsHandler(_ req: Request) throws -> Future<[PostItem]> {
-        _ = try req.requireAuthenticated(User.self)
-        
-        return PostItem.query(on: req).all()
+        try req.authorizedUser().flatMap { _ in
```
Comment (arkus7) [inline Sources/App/Controllers/PostController.swift]: I think this is not needed here 🤔 Going to remove it, as the user is already authorized in the middleware, no need to get them from DB here

Comment (0xLeif) [reply Sources/App/Controllers/PostController.swift]: Ya looks like it :+1:  Good catch

File: Sources/App/Controllers/UserController.swift [hunk 2]
```diff
@@ -37,37 +36,41 @@ class UserController: RouteCollection {
...
-            try req.authenticateSession(user)
-            return PublicUserResponse(id: user.id, email: user.email, social: user.social)
-        }
-        .map(to: HTTPResponse.self) { publicUser in
-            var response = HTTPResponse(status: .created)
-            try JSONEncoder().encode(publicUser, to: &response, on: req)
-            return response
+            let publicUser = PublicUserResponse(id: user.id, email: user.email, social: user.social)
+            return try authService.createAccessToken(for: user, on: req).map({
+                LoginResponse(token: $0, user: publicUser)
```
Comment (0xLeif) [inline Sources/App/Controllers/UserController.swift:59]: Nice :+1:

File: Sources/App/Services/AuthService.swift [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            .filter(\.token == refreshToken)
+            .first()
+            .unwrap(or: Abort(.unauthorized))
+        
+        return refreshTokenModel.flatMap { refreshTokenModel in
+            if refreshTokenModel.expiredAt > Date() {
+                return refreshTokenModel.user.get(on: request)
+                    .flatMap { user in
+                        return try self.createAccessToken(for: user, on: request)
+                }.then { accessToken in
```
Comment (0xLeif) [inline Sources/App/Services/AuthService.swift:55]: Cool `then` :eyes:

File: Sources/App/Utilities/JWT/JWTMiddleware.swift [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+//
+
+import Vapor
+import JWT
+
+class JWTMiddleware: Middleware {
+    
+    func respond(to request: Request, chainingTo next: Responder) throws -> EventLoopFuture<Response> {
+        let token = request.token
+        if !token.isEmpty {
```
Comment (0xLeif) [inline Sources/App/Utilities/JWT/JWTMiddleware.swift]: This feels more swifty to me... 
```
guard !token.isEmpty else {
      throw Abort(.unauthorized, reason: "No Access Token")
}
```

What do you think?

File: Sources/App/Utilities/JWT/TokenHelpers.swift [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+//
+//  Created by Arkadiusz Żmudzin on 13/03/2020.
+//
+
+import JWT
+
+class TokenHelpers {
+    /// Create payload for Access Token
+    fileprivate class func createPayload(from user: User) throws -> AccessTokenPayload {
+        if let id = user.id {
```
Comment (0xLeif) [inline Sources/App/Utilities/JWT/TokenHelpers.swift]: Same here. Would prefer a `guard let`

File: Sources/App/Utilities/JWT/TokenHelpers.swift [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    
+    /// Create Access Token for user
+    class func createAccessToken(from user: User) throws -> String {
+        let payload = try TokenHelpers.createPayload(from: user)
+        let header = JWTConfig.header
+        let signer = JWTConfig.signer
+        let jwt = JWT<AccessTokenPayload>(header: header, payload: payload)
+        let tokenData = try signer.sign(jwt)
+        
+        if let token = String(data: tokenData, encoding: .utf8) {
```
Comment (0xLeif) [inline Sources/App/Utilities/JWT/TokenHelpers.swift]: :eyes:

File: Sources/App/Utilities/Request+AuthorizedUser.swift [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+//
+//  Created by Arkadiusz Żmudzin on 13/03/2020.
+//
+
+import Vapor
+import JWT
+
+extension Request {
+    var token: String {
+        if let token = self.http.headers[.authorization].first?.replacingOccurrences(of: "Bearer ", with: "") {
```
Comment (0xLeif) [inline Sources/App/Utilities/Request+AuthorizedUser.swift]: `self.http.headers[.authorization].first?.replacingOccurrences(of: "Bearer ", with: "") ?? ""`

File: Sources/App/Utilities/JWT/JWTMiddleware.swift (hunk 4 above)
Comment (arkus7) [reply Sources/App/Utilities/JWT/JWTMiddleware.swift]: Sure, as you can clearly see, I forgot about `guard` syntax :D

File: Sources/App/Controllers/ImperialController.swift [hunk 8]
```diff
@@ -16,35 +16,37 @@ struct ImperialController: RouteCollection {
...
   func processGoogleLogin(request: Request, token: String) throws -> Future<ResponseEncodable> {
     return try Google.getUser(on: request).flatMap(to: ResponseEncodable.self) { userInfo in
       return User.query(on: request).filter(\.email == userInfo.email)
                  .first().flatMap(to: ResponseEncodable.self) { foundUser in
-        guard let existingUser = foundUser else {
+        guard let _ = foundUser else {
           let user = User(email: userInfo.email, password: UUID().uuidString)
           return user.save(on: request).map(to: ResponseEncodable.self) { user in
-            try request.authenticateSession(user)
+//            try request.authenticateSession(user)
```
Comment (ambid17) [inline Sources/App/Controllers/ImperialController.swift:27]: non-used comments?

File: Sources/App/Controllers/ImperialController.swift [hunk 9]
```diff
@@ -16,35 +16,37 @@ struct ImperialController: RouteCollection {
...
             return request.redirect(to: "/")
           }
         }
-        try request.authenticateSession(existingUser)
+//        try request.authenticateSession(existingUser)
```
Comment (ambid17) [inline Sources/App/Controllers/ImperialController.swift:31]: here too

File: Sources/App/Controllers/ImperialController.swift [hunk 10]
```diff
@@ -16,35 +16,37 @@ struct ImperialController: RouteCollection {
...
   func processGitHubLogin(request: Request, token: String) throws -> Future<ResponseEncodable> {
     return try GitHub.getUser(on: request).flatMap(to: ResponseEncodable.self) { userInfo in
       return User.query(on: request).filter(\.email == userInfo.login)
                  .first().flatMap(to: ResponseEncodable.self) { foundUser in
-        guard let existingUser = foundUser else {
+        guard let _ = foundUser else {
           let user = User(email: userInfo.login, password: UUID().uuidString)
           return user.save(on: request).map(to: ResponseEncodable.self) { user in
-            try request.authenticateSession(user)
+//            try request.authenticateSession(user)
```
Comment (ambid17) [inline Sources/App/Controllers/ImperialController.swift:45]: here

File: Sources/App/Controllers/ImperialController.swift [hunk 11]
```diff
@@ -16,35 +16,37 @@ struct ImperialController: RouteCollection {
...
             return request.redirect(to: "/")
           }
         }
-        try request.authenticateSession(existingUser)
+//        try request.authenticateSession(existingUser)
```
Comment (ambid17) [inline Sources/App/Controllers/ImperialController.swift:49]: aaand here

**Review by ambid17 — APPROVED** [review]
some extra comments, but otherwise, AMAZING work

File: Sources/App/Controllers/ImperialController.swift (hunk 8 above)
Comment (arkus7) [reply Sources/App/Controllers/ImperialController.swift:27]: @ambid17 I've added `TODO` comment on the function level. 
I left it here just to remember what to do when we'll be adjusting this later to handle Github/Google login :)

File: Sources/App/Utilities/JWT/JWTMiddleware.swift [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+//
+
+import Vapor
+import JWT
+
+class JWTMiddleware: Middleware {
+    
+    func respond(to request: Request, chainingTo next: Responder) throws -> EventLoopFuture<Response> {
+        let token = request.token
+        guard !token.isEmpty else {
```
Comment (0xLeif) [inline Sources/App/Utilities/JWT/JWTMiddleware.swift]: :+1:

File: Sources/App/Utilities/JWT/JWTMiddleware.swift [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+            throw Abort(.unauthorized, reason: "No Access Token")
+        }
```
Comment (0xLeif) [inline Sources/App/Utilities/JWT/JWTMiddleware.swift:17]: :+1:

File: Sources/App/Utilities/JWT/TokenHelpers.swift [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            return receivedJWT.payload
+        } catch {
+            throw JWTError.verificationFailed
+        }
+    }
+    
+    /// Generate new Refresh Token
+    class func createRefreshToken() -> String {
+        let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+        return String((0 ... 40).map { _ in letters.randomElement()! })
```
Comment (0xLeif) [inline Sources/App/Utilities/JWT/TokenHelpers.swift:73]: :exclamation: :eyes:

File: Tests/AppTests/Controllers/UserController/LoginTests.swift [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    let loginUri = "/api/login"
+    
+    static var app: Application!
+    static var connection: PostgreSQLConnection!
+    static var userService: UserService!
+    
+    let userEmail = "test@test.com"
+    let userPassword = "Password#1"
+    let invalidEmail = "invalid@email.com"
+    let invalidPassword = "InvalidPassword:("
```
Comment (0xLeif) [inline Tests/AppTests/Controllers/UserController/LoginTests.swift:23]: :(

**Review by 0xLeif — APPROVED** [review]

**Pull request merged**